### PR TITLE
ml water example

### PR DESCRIPTION
**Pull request by ChristianBeilschmidt**

### CODE REVIEW

**Comment by ChristianBeilschmidt** [comment]
pyarrow is not yet compatible with Python 3.13. Do we merge it anyway?

**Comment by ChristianBeilschmidt** [comment]
[ML mit Sentinel 2 + Improvements](https://trello.com/c/17AHdlOx/1734-ml-mit-sentinel-2-improvements)

File: geoengine/ml.py [hunk 1]
```diff
@@ -6,10 +6,14 @@
 import tempfile
 from typing import Protocol
 from dataclasses import dataclass
-from geoengine_openapi_client.models import MlModelMetadata, MlModel
+from onnx.reference import ReferenceEvaluator
```
Comment (michaelmattig) [inline geoengine/ml.py:9]: unrelated to this line: why did you add "zipmap: false" to the ml_pipeline.ipynb? it worked without it

File: geoengine/ml.py [hunk 2]
```diff
@@ -30,12 +34,19 @@ class MlModelConfig:
     description: str = "My Ml Model Description"
 
 
-def register_ml_model(onnx_model: SerializableModel,
+def register_ml_model(onnx_model: ModelProto,
```
Comment (michaelmattig) [inline geoengine/ml.py:37]: you can remove the SerializableModel class then I guess?

File: geoengine/ml.py [hunk 3]
```diff
@@ -56,3 +67,64 @@ def register_ml_model(onnx_model: SerializableModel,
...
+                                 f'expected type `{expected_type}`')
+
+    for domain in onnx_model.opset_import:
+        if domain.domain != '':
+            continue
+        if domain.version != 9:
+            raise InputException('Only ONNX models with opset version 9 are supported')
+
+    try:
+        model_ref = ReferenceEvaluator(onnx_model)
```
Comment (michaelmattig) [inline geoengine/ml.py:94]: I am not sure if this is the correct way of checking the model... zipMap works fine in our ONNX operator.

File: geoengine/ml.py [hunk 4]
```diff
@@ -56,3 +67,64 @@ def register_ml_model(onnx_model: SerializableModel,
...
+    except NotImplementedError as e:
+        if 'ZipMap' in str(e):
+            raise InputException(
+                'ZipMap is not supported in the model. Consider setting `options={"zipmap": False}`'
+            ) from e
+        raise e  # just re-raise
+
+    if len(model_ref.input_types) != 1:
+        raise InputException('Models with multiple inputs are not supported')
```
Comment (michaelmattig) [inline geoengine/ml.py:103]: shouldn't this be checked in the OnnxOperator class?

File: geoengine/ml.py [hunk 5]
```diff
@@ -6,10 +6,14 @@
 import tempfile
 from typing import Protocol
 from dataclasses import dataclass
-from geoengine_openapi_client.models import MlModelMetadata, MlModel
+from onnx.reference import ReferenceEvaluator
+from onnx import TypeProto, TensorProto, ModelProto
```
Comment (michaelmattig) [inline geoengine/ml.py:10]: unrelated to this line: is it possible to show the interactive point labeling tool in the jupyter notebook (on github)? Right now in VSCode I get an error and in Githug there is just text.

If it's not possible maybe add a text that it's not shown here?

File: geoengine/ml.py (hunk 1 above)
Comment (ChristianBeilschmidt) [reply geoengine/ml.py:9]: (a) It seems to be not a favorable option: https://onnx.ai/sklearn-onnx/versions/v1.16.0/auto_examples/plot_convert_zipmap.html#let-s-compare-prediction-time
(b) it does not work with the ReferenceEvaluator

File: geoengine/ml.py (hunk 5 above)
Comment (ChristianBeilschmidt) [reply geoengine/ml.py:10]: ipywidgets are not shown on GitHub currently.

I tried to save it using a jupyter notebook viewer and https://ipywidgets.readthedocs.io/en/stable/embedding.html#embedding-widgets-in-the-sphinx-html-documentation

[c2b2bd6](https://github.com/geo-engine/geoengine-python/pull/196/commits/c2b2bd6d696f11ea72a1bb1e8e4cc9c0511c58fa)

File: geoengine/ml.py (hunk 2 above)
Comment (ChristianBeilschmidt) [reply geoengine/ml.py:29]: [9bea96e](https://github.com/geo-engine/geoengine-python/pull/196/commits/9bea96e2b1118518423c66c7c499bacefe503a83)

File: geoengine/ml.py (hunk 5 above)
Comment (ChristianBeilschmidt) [reply geoengine/ml.py:10]: Does not seem to do much :shrug:

File: geoengine/ml.py (hunk 2 above)
Comment (ChristianBeilschmidt) [reply geoengine/ml.py:28]: [152421b](https://github.com/geo-engine/geoengine-python/pull/196/commits/152421b99471daf09897d1c3e29dd86be33f75db)

File: geoengine/ml.py (hunk 3 above)
Comment (ChristianBeilschmidt) [reply geoengine/ml.py:85]: https://github.com/onnx/sklearn-onnx/issues/218

ZipMap is debatable. Is it important?

File: geoengine/ml.py (hunk 5 above)
Comment (michaelmattig) [reply geoengine/ml.py:9]: ok then maybe write a line why nothing is visible or just resolve this comment if you dont want to do that.

**Comment by michaelmattig** [comment]
> pyarrow is not yet compatible with Python 3.13. Do we merge it anyway?

I'd say we remove Python 3.13. and add a task to add it later.

File: geoengine/ml.py (hunk 5 above)
Comment (ChristianBeilschmidt) [reply geoengine/ml.py:8]: [77beb8a](https://github.com/geo-engine/geoengine-python/pull/196/commits/77beb8a47e17a691de20d06e290dc0bb2083d537)

File: geoengine/ml.py (hunk 1 above)
Comment (ChristianBeilschmidt) [reply geoengine/ml.py]: [0df8a27](https://github.com/geo-engine/geoengine-python/pull/196/commits/0df8a27321e7cbebf477775d1741b20f61adc03b)

**Review by michaelmattig — APPROVED** [review]
